hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/associations/TokenAssociationsDumpUtils.java

```diff
@@ -0,0 +1,150 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class TokenAssociationsDumpUtils {
+    public static void dumpModTokenRelations(
+            @NonNull final Path path,
+            @NonNull
+                    final VirtualMap<OnDiskKey<com.hedera.hapi.node.base.TokenAssociation>, OnDiskValue<TokenRelation>>
+                            associations,
+            @NonNull final DumpCheckpoint checkpoint) {
+        try (@NonNull final var writer = new Writer(path)) {
+            final var dumpableTokenRelations = gatherTokenRelations(
+                    associations, key -> TokenAssociationId.fromMod(key.getKey()), TokenAssociation::fromMod);
+            reportOnUniques(writer, dumpableTokenRelations);
+            System.out.printf(
+                    "=== mod token associations report is %d bytes at checkpoint %s%n",
+                    writer.getSize(), checkpoint.name());
+        }
+    }
+
+    public static void dumpMonoTokenRelations(
+            @NonNull final Path path,
+            @NonNull final VirtualMap<OnDiskKey<OnDiskTokenRel>, OnDiskTokenRel> associations,
+            @NonNull final DumpCheckpoint checkpoint) {
+        try (@NonNull final var writer = new Writer(path)) {
+            final var dumpableTokenRelations =
+                    gatherTokenRelations(associations, TokenAssociationId::fromMono, TokenAssociation::fromMono);
+            reportOnUniques(writer, dumpableTokenRelations);
+            System.out.printf(
+                    "=== mono token associations report is %d bytes at checkpoint %s%n",
+                    writer.getSize(), checkpoint.name());
+        }
+    }
+
+    @NonNull
+    private static <K extends VirtualKey, V extends VirtualValue>
+            Map<TokenAssociationId, TokenAssociation> gatherTokenRelations(
+                    @NonNull final VirtualMap<K, V> source,
+                    @NonNull final Function<K, TokenAssociationId> keyMapper,
+                    @NonNull final Function<V, TokenAssociation> valueMapper) {
+        final var r = new HashMap<TokenAssociationId, TokenAssociation>();
+        final var threadCount = 8;
+        final var tokenAssociations = new ConcurrentLinkedQueue<Pair<TokenAssociationId, TokenAssociation>>();
+        try {
+            VirtualMapLike.from(source)
+                    .extractVirtualMapData(
+                            getStaticThreadManager(),
+                            p -> tokenAssociations.add(
+                                    Pair.of(keyMapper.apply(p.left()), valueMapper.apply(p.right()))),
+                            threadCount);
+        } catch (final InterruptedException ex) {
+            System.err.println("*** Traversal of token associations virtual map interrupted!");
+            Thread.currentThread().interrupt();
+        }
+        while (!tokenAssociations.isEmpty()) {
```

Comment:
Nothing wrong here but you could consider using `ConcurrentLinkedQueue.forEach`.